src/test/lib/optimizer/strategy/join_to_predicate_rewrite_rule_test.cpp

```diff
@@ -0,0 +1,83 @@
+#include "logical_query_plan/change_meta_table_node.hpp"
+#include "logical_query_plan/delete_node.hpp"
+#include "logical_query_plan/export_node.hpp"
+#include "logical_query_plan/insert_node.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/mock_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "logical_query_plan/sort_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "logical_query_plan/update_node.hpp"
+#include "optimizer/strategy/join_to_predicate_rewrite_rule.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+class JoinToPredicateRewriteRuleTest : public StrategyBaseTest {
+ public:
+  void SetUp() override {
+    node_a = MockNode::make(
+        MockNode::ColumnDefinitions{{DataType::Int, "a"}, {DataType::Int, "b"}, {DataType::Int, "c"}}, "a");
+    node_b = MockNode::make(
+        MockNode::ColumnDefinitions{{DataType::Int, "u"}, {DataType::Int, "v"}, {DataType::Int, "w"}}, "b");
+
+    a = node_a->get_column("a");
+    b = node_a->get_column("b");
+    c = node_a->get_column("c");
+    u = node_b->get_column("u");
+    v = node_b->get_column("v");
+    w = node_b->get_column("w");
+
+    auto key_constraints = std::vector<TableKeyConstraint>();
+
+    key_constraints.push_back(TableKeyConstraint({u->original_column_id}, KeyConstraintType::UNIQUE));
+    key_constraints.push_back(TableKeyConstraint({v->original_column_id}, KeyConstraintType::UNIQUE));
+
+    node_b->set_key_constraints(key_constraints);
+
+    rule = std::make_shared<JoinToPredicateRewriteRule>();
+  }
+
+  std::shared_ptr<JoinToPredicateRewriteRule> rule;
+  std::shared_ptr<MockNode> node_a, node_b;
+  std::shared_ptr<LQPColumnExpression> a, b, c, u, v, w;
+};
+
+TEST_F(JoinToPredicateRewriteRuleTest, SimplePredicateSemiJoin) {
+  auto lqp = std::shared_ptr<AbstractLQPNode>{};
+
+  // clang-format off
+  lqp =
+  ProjectionNode::make(std::vector<std::shared_ptr<AbstractExpression>>{b},
+    JoinNode::make(JoinMode::Semi, equals_(a, u),
+      node_a,
+      PredicateNode::make(equals_(v, 0), node_b)));
+
+  const auto actual_lqp = apply_rule(rule, lqp);
+
+  const auto subquery = ProjectionNode::make(std::vector<std::shared_ptr<AbstractExpression>>{u},
+    PredicateNode::make(equals_(v, 0), node_b));
+  const auto param_ids = std::vector<ParameterID>{};
+  const auto param_expressions = std::vector<std::shared_ptr<AbstractExpression>>{};
+
+  const auto expected_lqp =
+  ProjectionNode::make(std::vector<std::shared_ptr<AbstractExpression>>{b},
+    PredicateNode::make(equals_(a, std::make_shared<LQPSubqueryExpression>(subquery, param_ids, param_expressions)),
+      node_a));
+  // clang-format on
+
+  // We explicitly need to empty the mock node's key constraints here, otherwise, the node's equality comparison doesn't work
```

Comment:
Pick here: Apply your correction to `MockNode::_on_equals()` that fixes the comparison.

Reply:
I added the correction to the equals method of `MockNode`. Unfortunately, the test still fails, although the two compared LQPs are equal in value. I checked and there still seems to be a problem with the functional dependencies of mock nodes. Both now have the same 2 functional dependencies, when I print them, they look exactly the same, but the `==` check returns `false`. The mock nodes used in the two trees have an equal schema, but are not identical. Is it possible that in the comparison of mock nodes or (column) expressions, there is an identity check somewhere?

Reply:
```c++
bool MockNode::_on_shallow_equals(const AbstractLQPNode& rhs, const LQPNodeMapping& node_mapping) const {
  const auto& mock_node = static_cast<const MockNode&>(rhs);
  return _column_definitions == mock_node._column_definitions && _pruned_column_ids == mock_node._pruned_column_ids &&
         mock_node.name == name && mock_node.key_constraints() == _table_key_constraints &&
         mock_node._functional_dependencies == _functional_dependencies;
}
```

works for me, i.e,
```c++
  auto node_a = MockNode::make(MockNode::ColumnDefinitions{{DataType::Int, "a"}, { DataType::Long, "b"}});
  auto node_b = MockNode::make(MockNode::ColumnDefinitions{{DataType::Int, "a"}, { DataType::Long, "b"}});

  node_a->set_key_constraints({TableKeyConstraint{std::set<ColumnID>{ColumnID{0}}, KeyConstraintType::UNIQUE}});
  node_b->set_key_constraints({TableKeyConstraint{std::set<ColumnID>{ColumnID{0}}, KeyConstraintType::UNIQUE}});

  if (*node_a == *node_b) {
    std::cout << "node_a equals node_b" << std::endl;
  }
```

prints
```
node_a equals node_b
```

When you compare `functional_dependencies()` rather than `_functional_dependencies`, the `MockNode` builds FDs from unique constraints based on its output expressions, and these FDs will contain column expressions that are not from the same nodes (i.e., not equal). However, we check the set key constraints, and this should be fine.

Reply:
Thanks, taking the member variable instead of the function fixed the tests.